File: src/main/java/no/nav/modiaeventdistribution/kafka/KafkaConsumer.java
package no.nav.modiaeventdistribution.kafka;

import lombok.extern.slf4j.Slf4j;
import no.nav.apiapp.selftest.Helsesjekk;
import no.nav.apiapp.selftest.HelsesjekkMetadata;
import no.nav.validation.ValidationUtils;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.common.security.plain.PlainLoginModule;
import org.apache.kafka.common.serialization.StringDeserializer;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;
import java.util.function.BiConsumer;

import static java.util.stream.Collectors.joining;
import static no.nav.sbl.util.EnvironmentUtils.requireEnvironmentName;
import static org.apache.kafka.clients.CommonClientConfigs.SECURITY_PROTOCOL_CONFIG;
import static org.apache.kafka.clients.consumer.ConsumerConfig.*;
import static org.apache.kafka.common.config.SaslConfigs.SASL_JAAS_CONFIG;

@Slf4j
public class KafkaConsumer implements Helsesjekk {

    private static final long ERROR_GRACE_PERIOD = Duration.ofMinutes(5).toMillis();
    public static final Duration POLL_TIMEOUT = Duration.ofSeconds(5);
    private final HelsesjekkMetadata helsesjekkMetadata;
    private final org.apache.kafka.clients.consumer.KafkaConsumer<String, String> consumer;
    private final String topicName;
    private final BiConsumer<String, String> handler;

    private volatile boolean running;
    private volatile long lastErrorTimestamp;
    private volatile Throwable lastError;

    public KafkaConsumer(KafkaConsumerConfig kafkaConsumerConfig) {
        ValidationUtils.validate(kafkaConsumerConfig);

        this.topicName = kafkaConsumerConfig.topicName + "-" + requireEnvironmentName();
        this.handler = kafkaConsumerConfig.handler;

        String bootstrapServers = kafkaConsumerConfig.bootstrapServers.stream().map(s -> String.format("%s:%s", s.host, s.port)).collect(joining(","));
        String groupId = kafkaConsumerConfig.groupId;

        this.helsesjekkMetadata = new HelsesjekkMetadata(
                "kafka-consumer-" + topicName,
                bootstrapServers,
                String.format("consuming kafka topic: %s with groupId %s", topicName, groupId),
                false
        );

        Properties props = new Properties();
        props.put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(GROUP_ID_CONFIG, groupId);
        props.put(KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());

        props.put(SECURITY_PROTOCOL_CONFIG, "SASL_SSL");
        props.put("sasl.mechanism", "PLAIN");
        props.put(SASL_JAAS_CONFIG, String.format("%s required username=\"%s\" password=\"%s\";",
                PlainLoginModule.class.getName(),
                kafkaConsumerConfig.username,
                kafkaConsumerConfig.password
        ));

        consumer = new org.apache.kafka.clients.consumer.KafkaConsumer<>(props);
    }

    public void start() {
        running = true;

        log.info("starting kafka consumer for topic {}", topicName);
        consumer.subscribe(Collections.singleton(topicName));

        Thread thread = new Thread(this::run);
        thread.setName("consumer-" + topicName);
        thread.setDaemon(true);
        thread.start();
    }

    private void run() {
        while (running) {
            try {
                ConsumerRecords<String, String> records = consumer.poll(POLL_TIMEOUT);
                for (ConsumerRecord<String, String> record : records) {
                    handler.accept(record.value(), record.key());
                }
                consumer.commitAsync();
            } catch (Throwable e) {
                log.error(e.getMessage(), e);
                lastErrorTimestamp = System.currentTimeMillis();
                lastError = e;
            }
        }
    }

    public void stop() {
        running = false;
    }

    @Override
    public void helsesjekk() throws Throwable {
        if (!running) {
            throw new IllegalStateException("consumer is not running");
        }

        if (lastErrorTimestamp + ERROR_GRACE_PERIOD > System.currentTimeMillis()){
            throw lastError;
        }
    }

    @Override
    public HelsesjekkMetadata getMetadata() {
        return helsesjekkMetadata;
    }

}

File: src/main/kotlin/no/nav/modiaeventdistribution/kafka/KafkaConsumer.kt
package no.nav.modiaeventdistribution.kafka

import kotlinx.coroutines.runBlocking
import no.nav.common.health.HealthCheckResult
import no.nav.common.health.selftest.SelfTestCheck
import no.nav.modiaeventdistribution.Config
import no.nav.modiaeventdistribution.infrastructur.HealthCheckAware
import no.nav.modiaeventdistribution.log
import org.apache.kafka.clients.CommonClientConfigs
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecords
import org.apache.kafka.common.config.SaslConfigs
import org.apache.kafka.common.security.plain.PlainLoginModule
import org.apache.kafka.common.serialization.StringDeserializer
import java.time.Duration
import java.util.*

data class BootstrapServer internal constructor(
        val host: String,
        val port: Int
) {
    companion object {
        fun parse(configString: String): List<BootstrapServer> {
            return configString.split(",")
                    .map {
                        val (host, port) = it.split(":")
                        BootstrapServer(host, port.toInt())
                    }
        }
    }
}

data class KafkaConsumers(val consumers: List<KafkaConsumer>) {
    fun start() = consumers.forEach(KafkaConsumer::start)
    fun stop() = consumers.forEach(KafkaConsumer::stop)
}

internal val ERROR_GRACE_PERIODE = Duration.ofMinutes(5).toMillis()
internal val POLL_TIMEOUT = Duration.ofSeconds(5)

class KafkaConsumer(
        config: Config,
        topicName: String,
        groupId: String,
        bootstrapServers: List<BootstrapServer>,
        username: String,
        password: String,
        private val handler: suspend (key: String?, value: String?) -> Unit
) : HealthCheckAware {
    private val healthCheckData: SelfTestCheck
    private val consumer: org.apache.kafka.clients.consumer.KafkaConsumer<String, String>
    private val topicNameWithEnv: String

    @Volatile
    private var running = false

    @Volatile
    private var lastErrorTimestamp: Long = 0

    @Volatile
    private var lastError: Throwable? = null

    init {
        val bootstrapServersString = bootstrapServers.joinToString(",") { (host, port) -> "$host:$port" }
        this.topicNameWithEnv = "${topicName}-${config.appEnvironment}"
        this.healthCheckData = SelfTestCheck(
                """
                    topic: ${this.topicNameWithEnv}
                    groupId: ${groupId}
                    server: $bootstrapServersString
                """.trimIndent()
                , false
        ) { this.checkHealth() }


        val props = Properties()
        props[ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG] = bootstrapServersString
        props[ConsumerConfig.GROUP_ID_CONFIG] = groupId
        props[ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java.name
        props[ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java.name
        props[CommonClientConfigs.SECURITY_PROTOCOL_CONFIG] = "SASL_SSL"
        props["sasl.mechanism"] = "PLAIN"
        props[SaslConfigs.SASL_JAAS_CONFIG] = String.format("%s required username=\"%s\" password=\"%s\";",
                PlainLoginModule::class.java.name,
                username,
                password
        )
        this.consumer = org.apache.kafka.clients.consumer.KafkaConsumer(props)
    }

    fun start() {
        this.running = true
        log.info("starting kafka consumer for topic {}", topicNameWithEnv)
        this.consumer.subscribe(setOf(topicNameWithEnv))

        val thread = Thread(Runnable { this.run() })
        thread.name = "consumer-$topicNameWithEnv"
        thread.isDaemon = true
        thread.start()
    }

    fun stop() {
        this.running = false
    }

    fun run() {
        while (this.running) {
            try {
                val records: ConsumerRecords<String?, String?> = this.consumer.poll(POLL_TIMEOUT)
                runBlocking {
                    log.info("Received kafka-messages: ${records.count()}")
                    for (record in records) {
                        handler(record.key(), record.value())
                    }
                }
                this.consumer.commitAsync()
            } catch (e: Throwable) {
                log.error(e.message, e)
                this.lastErrorTimestamp = System.currentTimeMillis()
                this.lastError = e
            }
        }
    }

    override fun getHealthCheck(): SelfTestCheck = this.healthCheckData

    private fun checkHealth(): HealthCheckResult {
        if (!this.running) {
            return HealthCheckResult.unhealthy("consumer is not running")
        }
        if (this.lastErrorTimestamp + ERROR_GRACE_PERIODE > System.currentTimeMillis()) {
            return HealthCheckResult.unhealthy(lastError as Throwable)
        }
        return HealthCheckResult.healthy()
    }
}
